### ISSUE TITLE
Running behind ARR causes redirect loop for [RequireHttps] with default settings

### ISSUE THREAD

**Issue by mikes-gh** [aspnet/IISIntegration#111]
Though I'd add this to hopefully help someone else.

This took me a few hours to resolve.

A default rewrite rule in ARR does not insert X-FORWARDED-PROTO into the request.

Due to recent changes here the middleware only accepts sets of values

https://github.com/aspnet/BasicMiddleware/blob/dev/src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs

This causes

```
Parameter count mismatch between X-Forwarded-For and X-Forwarded-Proto.
```

Fortunately by reverse proxy is under my control.
You need to add 

HTTP_X_FORWARDED_PROTO = https

To the server Variables section of the inbound rewrite rule.in ARR.

Once this is done you also need to add addtional middleware to Configure in Startup.cs

```
ForwardedHeadersOptions options = new ForwardedHeadersOptions
            {
                ForwardedHeaders = Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedFor | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedProto,
                ForwardLimit = 2
            };
            options.KnownProxies.Add(IPAddress.Parse("192.168.1.58"));
app.UseForwardedHeaders(options);
```

Where 192.168.1.58 is my ARR server othewise I get

```
Unknown proxy: 192.168.1.58
```

I see that UseIIS() now sets up its own ForwardedHeadersOptions.  Can I add to the options setup by this extension method?

I wonder if their is an easier way to get this all working?

**Comment by Tratcher** [comment]
You can configure the ForwardedHeaderOptions in Startup.ConfigureServices like this:
```
            services.Configure<ForwardedHeadersOptions>(options =>
            {
            });
```
https://github.com/aspnet/IISIntegration/blob/dev/samples/IISSample/Startup.cs#L19

**Comment by mikes-gh** [comment]
Thanks @Tratcher that worked great.
I think the defaults set below should belong in UseIIS()

https://github.com/aspnet/BasicMiddleware/blob/dev/src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs

You may want setup ForwardHeaderOptions for scenarios where the loopback address is not applicable.

**Comment by escobar5** [comment]
Hi @mikes-gh, I'm having the same problem and I'm new to ARR, can you tell me where you put HTTP_X_FORWARDED_PROTO = https?

Also, I have my ARR and sites in the same server, do I need to configure the IP of the server on KnownProxies?

Thanks

**Comment by mikes-gh** [comment]
@escobar5 I'll post a screen shot monday.

**Comment by mikes-gh** [comment]
@escobar5 

Add it to your server variables as below

![rewrite rule](https://cloud.githubusercontent.com/assets/16208742/14420128/1ac72554-ffc2-11e5-9ec4-abb463ddc28a.png)

**Comment by escobar5** [comment]
thanks @mikes-gh, I'm going to try

**Comment by 1amirjalai** [comment]
I have an issue regarding x-forward-proto any help would be appreciated

i have whatever has been said on this issue thread but nothing is happening

my issue on stackoverflow

http://stackoverflow.com/questions/43492923/asp-net-core-parameter-count-mismatch-between-x-forwarded-for-and-x-forwarded-pr

**Comment by Tratcher** [comment]
See RequireHeaderSymmetry https://github.com/aspnet/BasicMiddleware/blob/6c5f812d77d2e43b0a7c49a50d238f51a9eb665f/src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs#L74

**Comment by 1amirjalai** [comment]
@Tratcher  I'm going to make RequireHeaderSymmetry  false

**Comment by 1amirjalai** [comment]
I made RequireHeaderSymmetry false but nothing happens still i get the error. i'm going to read the code asap

**Comment by mikes-gh** [comment]
Here is my working config

In ConfigureServices
```
services.Configure<ForwardedHeadersOptions>(options =>
            {
                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
                {
                    options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;
                    options.RequireHeaderSymmetry = false;
                }

                options.ForwardLimit = 2;
                options.KnownProxies.Add(IPAddress.Parse(Configuration["ReverseProxy"]));
            });
```

In   Configure
```
app.UseForwardedHeaders();
```

Hope this helps

**Comment by 1amirjalai** [comment]
no i still get 
Parameter count mismatch between X-Forwarded-For and X-Forwarded-Proto.

SourceContext
Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersMiddleware

@mikes-gh my config is exactly like yours

**Comment by armitagemderivitec** [comment]
@1amirjalai did you find a solution to this. I am using AWS Cloudfront and experienced the same issue because the x-forward-proto header is removed. Any insights you've had would be appreciated. 

Thanks

**Comment by 1amirjalai** [comment]
@armitagemderivitec 

the answer is  [here ](https://stackoverflow.com/questions/43492923/asp-net-core-parameter-count-mismatch-between-x-forwarded-for-and-x-forwarded-pr/44632440#44632440) at stackoverflow 


I found a solution just add this code on you configureservices

```
            services.Configure<ForwardedHeadersOptions>(options =>
            {
                options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;
                options.RequireHeaderSymmetry = false;
                options.ForwardLimit = 2;
                options.KnownProxies.Add(IPAddress.Parse("xxx.xxx.xxx.xxx"));
                options.KnownProxies.Add(IPAddress.Parse("xxx.xxx.xxx.xxx"));
                options.KnownProxies.Add(IPAddress.Parse("xxx.xxx.xxx.xxx"));
              
            });

```
and the on your basecontroller or somewhere else use this code for getting real ip address if you are behind cloudflare and arr


```
      public string GetRealIpAddress()
        {
            StringValues connectionIp;
            var conncetionIpAvailable = Request.Headers.TryGetValue("Cf-Connecting-Ip", out connectionIp);
            if (conncetionIpAvailable)
            {
                return connectionIp.ToString().Trim();
            }
            else
            {
                var xForwardFor = Request.Headers.GetCommaSeparatedValues("X-Forwarded-For");
                string ipList = "";
                foreach (var x in xForwardFor)
                {
                    ipList = ipList.Trim() + " , " + x.Trim();
                }

                if (!string.IsNullOrEmpty(ipList.Trim()))
                {
                    return ipList.Trim();
                }
                else
                {
                    return "xxx.xxx.xxx.xxx";
                }
            }
        }
```

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
